README.md
colorize [![Gem Version](https://badge.fury.io/rb/colorize.svg)](http://badge.fury.io/rb/colorize) [![Ruby Gem](https://github.com/fazibear/colorize/actions/workflows/release.yml/badge.svg)](https://github.com/fazibear/colorize/actions/workflows/release.yml) [![Code Climate](https://codeclimate.com/github/fazibear/colorize/badges/gpa.svg)](https://codeclimate.com/github/fazibear/colorize) [![Test Coverage](https://codeclimate.com/github/fazibear/colorize/badges/coverage.svg)](https://codeclimate.com/github/fazibear/colorize)
========

Ruby gem for colorizing text using ANSI escape sequences.
Extends `String` class or add a `ColorizedString` with methods to set the text color, background color and text effects.

modes
-----

* `require 'colorize'` - Extends String class
* `require 'colorized_string'` - add ColorizedString class

features
--------

* change string color
* change string background
* change string effect
* chain methods
* display color samples
* disable colorization
* prevent colors

usage
-----

```ruby
require 'colorize'

String.colors                       # return array of all possible colors names
String.modes                        # return array of all possible modes
String.color_samples                # displays color samples in all combinations
String.disable_colorization         # check if colorization is disabled

# disable colorization (false by default)

String.disable_colorization = false # enable colorization
String.disable_colorization false   # enable colorization
String.disable_colorization = true  # disable colorization
String.disable_colorization true    # disable colorization

# prevent colors (false by default)

String.prevent_colors = false # override current string colors
String.prevent_colors false   # override current string colors
String.prevent_colors = true  # don't colorize colorized strings
String.prevent_colors true    # don't colorize colorized strings

# adding aliases (:gray and :grey added by default)

String.add_color_alias(:niebieski, :blue)
String.add_color_alias(:zielony => :green) 
String.add_color_alias(czarny: :black)
String.add_color_alias(czerwony: :red, granatowy: :blue)

# examaples

puts "This is blue".colorize(:blue)
puts "This is light blue".colorize(:light_blue)
puts "This is also blue".colorize(:color => :blue)
puts "This is bold green".colorize(:color => :green, :mode => :bold)
puts "This is light blue with red background".colorize(:color => :light_blue, :background => :red)
puts "This is light blue with red background".colorize(:light_blue ).colorize( :background => :red)
puts "This is blue text on red".blue.on_red
puts "This is red on blue".colorize(:red).on_blue
puts "This is red on blue and underline".colorize(:red).on_blue.underline
puts "This is blinking blue text on red".blue.on_red.blink
puts "This is uncolorized".blue.on_red.uncolorize
```

```ruby
require 'colorized_string'

ColorizedString.colors                       # return array of all possible colors names
ColorizedString.modes                        # return array of all possible modes
ColorizedString.color_samples                # displays color samples in all combinations
ColorizedString.disable_colorization         # check if colorization is disabled

# disable colorization (false by default)

ColorizedString.disable_colorization = false # enable colorization
ColorizedString.disable_colorization false   # enable colorization
ColorizedString.disable_colorization = true  # disable colorization
ColorizedString.disable_colorization true    # disable colorization

# prevent colors (false by default)

ColorizedString.prevent_colors = false # override current string colors
ColorizedString.prevent_colors false   # override current string colors
ColorizedString.prevent_colors = true  # don't colorize colorized strings
ColorizedString.prevent_colors true    # don't colorize colorized strings

# adding aliases (:gray and :grey added by default)

ColorizedString.add_color_alias(:niebieski, :blue)
ColorizedString.add_color_alias(:zielony => :green) 
ColorizedString.add_color_alias(czarny: :black)
ColorizedString.add_color_alias(czerwony: :red, granatowy: :blue)

# examples

puts ColorizedString["This is blue"].colorize(:blue)
puts ColorizedString["This is light blue"].colorize(:light_blue)
puts ColorizedString["This is also blue"].colorize(:color => :blue)
puts ColorizedString["This is bold green"].colorize(:color => :green, :mode => :bold)
puts ColorizedString["This is light blue with red background"].colorize(:color => :light_blue, :background => :red)
puts ColorizedString["This is light blue with red background"].colorize(:light_blue ).colorize( :background => :red)
puts ColorizedString["This is blue text on red"].blue.on_red
puts ColorizedString["This is red on blue"].colorize(:red).on_blue
puts ColorizedString["This is red on blue and underline"].colorize(:red).on_blue.underline
puts ColorizedString["This is blinking blue text on red"].blue.on_red.blink
puts ColorizedString["This is uncolorized"].blue.on_red.uncolorize

puts ColorizedString.new("This is blue").blue
puts ColorizedString.new("This is light blue").colorize(:light_blue)
```

requirements
------------

* Win32/Console/ANSI (for Windows)

install
-------

* gem install colorize

*Note:* You may need to use sudo to install gems

thank you
---------

[![Become Patreon](https://c5.patreon.com/external/logo/become_a_patron_button.png)](https://www.patreon.com/bePatron?u=6912974)

license
-------

Copyright (C) 2007-2016 Michał Kalbarczyk

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


CHANGELOG
== 1.0.0 / 2023-03-23
  * add more modes
  * add aliases
  * grey and gray default aliases
  * update dependencies
  * add non-visible text delimiters
  * cleanup and refactor

== 0.8.1 / 2016-06-29
  * fix gemspec bug

== 0.8.0 / 2016-06-27
	* add ColorizedString class
	* update README file
	* add rubocop.yml and follow style
  * add italic mode
  * remove interpreter warrnings

== 0.7.7 / 2015-04-19
  * update gems

== 0.7.6 / 2015-04-18
  * fix bugs

== 0.7.5 / 2014-12-11
  * big code refactoring
  * disable_colorization feature added

== 0.7.4 / 2014-12-10
  * code cleanups

== 0.7.3 / 2014-05-19
  * fix new line maching

== 0.7.2 / 2014-04-08
  * tests cleanups
  * gem release date fixed

== 0.7.1 / 2014-04-02
  * handling wrong color values

== 0.7.0 / 2014-03-12
  * refactored to use regexp pattern matching
  * works with frozen strings
  * works with concatenated string

== 0.6.0 / 2013-09-25
  * code cleanups
  * bold mode fixed
  * STDOUT.isatty condition removed
  * doc updated
  * jeweler tasks removed
  * Rakefile updated

== 0.5.8 / 2009-12-02
  * code cleanups
  * removed 'ruby -w' warnings

== 0.5.7 / 2009-11-24
  * moved to github/gemcutter
  * 1.9 compatible

== 0.5.6 / 2007-08-27
  * rdocs Allison template
  * cleanups

== 0.5.5 / 2007-08-11
  * added effects methods
  * README files
  * new rake file

== 0.5.0 / 2007-04-22
  * initial rewritten version


test/test_helper.rb
require 'minitest/autorun'
require 'minitest/mock'

require File.dirname(__FILE__) + '/../lib/colorize'


test/test_colorize.rb
# frozen_string_literal: true

require 'minitest/autorun'
require "#{File.dirname(__FILE__)}/../lib/colorize"

class TestColorize < Minitest::Test
  def test_blue_symbol
    assert_equal "\001\033[0;34;49m\002This is blue\001\033[0m\002", 'This is blue'.colorize(:blue)
  end

  def test_incorrect_symbol
    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002",  'This is incorrect color'.colorize(:bold)
  end

  def test_incorrect_hash
    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", 'This is incorrect color'.colorize(:color => :bold)

    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", 'This is incorrect color'.colorize(:mode => :green)

    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", 'This is incorrect color'.colorize(:background => :bold)
  end

  def test_blue_hash
    assert_equal "\001\033[0;34;49m\002This is also blue\001\033[0m\002", 'This is also blue'.colorize(:color => :blue)
  end

  def test_light_blue_symbol
    assert_equal "\001\033[0;94;49m\002This is light blue\001\033[0m\002", 'This is light blue'.colorize(:light_blue)
  end

  def test_light_blue_with_red_background_hash
    assert_equal "\001\033[0;94;41m\002This is light blue with red background\001\033[0m\002", 'This is light blue with red background'.colorize(:color => :light_blue, :background => :red)
  end

  def test_light_blue_with_red_background_symbol_and_hash
    assert_equal "\001\033[0;94;41m\002This is light blue with red background\001\033[0m\002", 'This is light blue with red background'.colorize(:light_blue).colorize(:background => :red)
  end

  def test_blue_with_red_background_method
    assert_equal "\001\033[0;34;41m\002This is blue text on red\001\033[0m\002", 'This is blue text on red'.blue.on_red
  end

  def test_red_with_blue_background_symbol_and_method
    assert_equal "\001\033[0;31;44m\002This is red on blue\001\033[0m\002", 'This is red on blue'.colorize(:red).on_blue
  end

  def test_red_with_blue_background_and_underline_symbol_and_methods
    assert_equal "\001\033[4;31;44m\002This is red on blue and underline\001\033[0m\002", 'This is red on blue and underline'.colorize(:red).on_blue.underline
  end

  def test_blue_with_red_background_and_blink_methods
    assert_equal "\001\033[5;34;41m\002This is blue text on red\001\033[0m\002", 'This is blue text on red'.blue.on_red.blink
  end

  def test_uncolorize
    assert_equal 'This is uncolorized', 'This is uncolorized'.blue.on_red.uncolorize
  end

  def test_colorized?
    assert_predicate 'Red'.red, :colorized?
    refute_predicate 'Blue', :colorized?
    refute_predicate 'Green'.blue.green.uncolorize, :colorized?
  end

  def test_concatenated__colorize?
    assert_predicate "none #{'red'.red} none #{'blue'.blue} none", :colorized?
    refute_predicate "none #{'red'.red} none #{'blue'.blue} none".uncolorize, :colorized?
  end

  def test_concatenated_strings_on_green
    assert_equal "\001\033[0;39;42m\002none \001\033[0m\002\001\033[0;31;42m\002red\001\033[0m\002\001\033[0;39;42m\002 none \001\033[0m\002\001\033[0;34;42m\002blue\001\033[0m\002\001\033[0;39;42m\002 none\001\033[0m\002", "none #{'red'.red} none #{'blue'.blue} none".on_green
  end

  def test_concatenated_strings_uncolorize
    assert_equal 'none red none blue none', "none #{'red'.red} none #{'blue'.blue} none".uncolorize
  end

  def test_new_line
    assert_equal "\001\033[5;34;41m\002This is blue\ntext on red\001\033[0m\002", "This is blue\ntext on red".blue.on_red.blink
  end

  def test_disable_colorization_with_=
    String.disable_colorization = true

    assert String.disable_colorization

    String.disable_colorization = false
  end

  def test_disable_colorization_with_method
    String.disable_colorization true

    assert String.disable_colorization

    String.disable_colorization false
  end

  def test_string_disable_colorization_with_=
    String.disable_colorization = true

    assert String.disable_colorization

    assert_equal 'This is blue after disabling', 'This is blue after disabling'.blue

    String.disable_colorization = false

    assert_equal "\001\033[0;34;49m\002This is blue after enabling\001\033[0m\002", 'This is blue after enabling'.colorize(:blue)
  end

  def test_string_disable_colorization_with_method
    assert_equal "\001\033[0;34;49m\002This is blue before disabling\001\033[0m\002", 'This is blue before disabling'.colorize(:blue)

    String.disable_colorization true

    assert String.disable_colorization

    assert_equal 'This is blue after disabling', 'This is blue after disabling'.blue

    String.disable_colorization false

    assert_equal "\001\033[0;34;49m\002This is blue after enabling\001\033[0m\002", 'This is blue after enabling'.colorize(:blue)
  end

  def test_already_colored_string_with_one_value
    assert_equal 'This is red'.red, "\001\033[31m\002This is red\001\033[0m\002".red
  end

  def test_color_matrix_method
    assert_raises NoMethodError do
      String.color_matrix
    end
  end

  def test_color_samples_method
    assert_output do
      String.color_samples
    end
  end

  def test_grey_alias
    assert_equal 'This is grey'.grey, 'This is grey'.light_black
  end

  def test_gray_alias
    assert_equal 'This is gray'.gray, 'This is gray'.light_black
  end

  def test_add_color_alias
    String.add_color_alias(:extra_blue, :light_blue)

    assert_equal 'blue'.light_blue, 'blue'.extra_blue
    assert_equal 'blue'.on_light_blue, 'blue'.on_extra_blue
  end

  def test_add_color_alias_errors
    String.add_color_alias(:extra_red, :light_red)

    assert_raises ::Colorize::ColorAlreadyExist, 'Colorize: color :extra_red already exist!' do
      String.add_color_alias(:extra_red, :light_blue)
    end

    assert_raises ::Colorize::ColorDontExist, 'Colorize: color :light_color don\'t exist!' do
      String.add_color_alias(:extra_white, :light_color)
    end
  end

  def test_add_color_alias_with_single_hash
    String.add_color_alias(extra_green: :light_green)

    assert_equal 'example string'.light_green, 'example string'.extra_green
    assert_equal 'example string'.on_light_green, 'example string'.on_extra_green
  end

  def test_add_color_alias_with_single_hash_with_arrow
    String.add_color_alias(:extra_color => :gray)

    assert_equal 'example string'.gray, 'example string'.extra_color
    assert_equal 'example string'.on_gray, 'example string'.on_extra_color
  end

  def test_add_color_alias_with_multi_hash
    String.add_color_alias(extra_color_a: :gray, extra_color_b: :blue)

    assert_equal 'example string'.gray, 'example string'.extra_color_a
    assert_equal 'example string'.blue, 'example string'.extra_color_b
  end

  def test_add_color_alias_with_multi_hash_with_arrow
    String.add_color_alias(:extra_color_c => :gray, :extra_color_d => :blue)

    assert_equal 'example string'.gray, 'example string'.extra_color_c
    assert_equal 'example string'.on_blue, 'example string'.on_extra_color_d
  end

  def test_add_color_alias_with_multi_hash_mixed
    String.add_color_alias(extra_color_e: :gray, :extra_color_f => :blue)

    assert_equal 'example string'.gray, 'example string'.extra_color_e
    assert_equal 'example string'.on_blue, 'example string'.on_extra_color_f
  end

  def test_prevent_colors
    String.prevent_colors true

    assert String.prevent_colors
    assert_equal "#{'blue'.blue}#{'red'.red}#{'green'.green}", "#{'blue'.blue}red#{'green'.green}".red

    String.prevent_colors false
  end
end


test/test_colorized_string.rb
# frozen_string_literal: true

require 'minitest/autorun'
require "#{File.dirname(__FILE__)}/../lib/colorized_string"

class TestColorizedString < Minitest::Test
  def test_colorize_not_required
    refute_respond_to '', :colorize, 'Colorize should not be required. Only colorized_string'
  end

  def test_blue_symbol
    assert_equal "\001\033[0;34;49m\002This is blue\001\033[0m\002", ColorizedString['This is blue'].colorize(:blue)
  end

  def test_incorrect_symbol
    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", ColorizedString['This is incorrect color'].colorize(:bold)
  end

  def test_incorrect_hash
    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", ColorizedString['This is incorrect color'].colorize(:color => :bold)

    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", ColorizedString['This is incorrect color'].colorize(:mode => :green)

    assert_equal "\001\033[0;39;49m\002This is incorrect color\001\033[0m\002", ColorizedString['This is incorrect color'].colorize(:background => :bold)
  end

  def test_blue_hash
    assert_equal "\001\033[0;34;49m\002This is also blue\001\033[0m\002", ColorizedString['This is also blue'].colorize(:color => :blue)
  end

  def test_light_blue_symbol
    assert_equal "\001\033[0;94;49m\002This is light blue\001\033[0m\002", ColorizedString['This is light blue'].colorize(:light_blue)
  end

  def test_light_blue_with_red_background_hash
    assert_equal "\001\033[0;94;41m\002This is light blue with red background\001\033[0m\002", ColorizedString['This is light blue with red background'].colorize(:color => :light_blue, :background => :red)
  end

  def test_light_blue_with_red_background_symbol_and_hash
    assert_equal "\001\033[0;94;41m\002This is light blue with red background\001\033[0m\002", ColorizedString['This is light blue with red background'].colorize(:light_blue).colorize(:background => :red)
  end

  def test_blue_with_red_background_method
    assert_equal "\001\033[0;34;41m\002This is blue text on red\001\033[0m\002", ColorizedString['This is blue text on red'].blue.on_red
  end

  def test_red_with_blue_background_symbol_and_method
    assert_equal "\001\033[0;31;44m\002This is red on blue\001\033[0m\002", ColorizedString['This is red on blue'].colorize(:red).on_blue
  end

  def test_red_with_blue_background_and_underline_sumbol_and_methods
    assert_equal "\001\033[4;31;44m\002This is red on blue and underline\001\033[0m\002", ColorizedString['This is red on blue and underline'].colorize(:red).on_blue.underline
  end

  def test_blue_with_red_background_and_blink_methods
    assert_equal "\001\033[5;34;41m\002This is blue text on red\001\033[0m\002", ColorizedString['This is blue text on red'].blue.on_red.blink
  end

  def test_uncolorize
    assert_equal 'This is uncolorized', ColorizedString['This is uncolorized'].blue.on_red.uncolorize
  end

  def test_colorized?
    assert_predicate ColorizedString['Red'].red, :colorized?
    refute_predicate ColorizedString['Blue'], :colorized?
    refute_predicate ColorizedString['Green'].blue.green.uncolorize, :colorized?
  end

  def test_concatenated__colorize?
    assert_predicate ColorizedString["none #{ColorizedString['red'].red} none #{ColorizedString['blue'].blue} none"], :colorized?
    refute_predicate ColorizedString["none #{ColorizedString['red'].red} none #{ColorizedString['blue'].blue} none"].uncolorize, :colorized?
  end

  def test_concatenated_strings_on_green
    assert_equal "\001\033[0;39;42m\002none \001\033[0m\002\001\033[0;31;42m\002red\001\033[0m\002\001\033[0;39;42m\002 none \001\033[0m\002\001\033[0;34;42m\002blue\001\033[0m\002\001\033[0;39;42m\002 none\001\033[0m\002", ColorizedString["none #{ColorizedString['red'].red} none #{ColorizedString['blue'].blue} none"].on_green
  end

  def test_concatenated_strings_uncolorize
    assert_equal 'none red none blue none', ColorizedString["none #{ColorizedString['red'].red} none #{ColorizedString['blue'].blue} none"].uncolorize
  end

  def test_frozen_strings
    assert_equal "\001\033[5;34;41m\002This is blue text on red\001\033[0m\002", ColorizedString['This is blue text on red'].freeze.blue.on_red.blink
  end

  def test_new_line
    assert_equal "\001\033[5;34;41m\002This is blue\ntext on red\001\033[0m\002", ColorizedString["This is blue\ntext on red"].freeze.blue.on_red.blink
  end

  def test_disable_colorization_with_=
    ColorizedString.disable_colorization = true

    assert ColorizedString.disable_colorization

    ColorizedString.disable_colorization = false
  end

  def test_disable_colorization_with_method
    ColorizedString.disable_colorization true

    assert ColorizedString.disable_colorization

    ColorizedString.disable_colorization false
  end

  def test_string_disable_colorization_with_=
    ColorizedString.disable_colorization = true

    assert ColorizedString.disable_colorization

    assert_equal 'This is blue after disabling', ColorizedString['This is blue after disabling'].blue

    ColorizedString.disable_colorization = false

    assert_equal "\001\033[0;34;49m\002This is blue after enabling\001\033[0m\002", ColorizedString['This is blue after enabling'].colorize(:blue)
  end

  def test_string_disable_colorization_with_method
    assert_equal "\001\033[0;34;49m\002This is blue before disabling\001\033[0m\002", ColorizedString['This is blue before disabling'].colorize(:blue)

    ColorizedString.disable_colorization true

    assert ColorizedString.disable_colorization

    assert_equal 'This is blue after disabling', ColorizedString['This is blue after disabling'].blue

    ColorizedString.disable_colorization false

    assert_equal "\001\033[0;34;49m\002This is blue after enabling\001\033[0m\002", ColorizedString['This is blue after enabling'].colorize(:blue)
  end

  def test_already_colored_string_with_one_value
    assert_equal ColorizedString['This is red'].red, ColorizedString["\001\033[31m\002This is red\001\033[0m\002"].red
  end

  def test_color_matrix_method
    assert_raises NoMethodError do
      ColorizedString.color_matrix
    end
  end

  def test_color_samples_method
    assert_output do
      ColorizedString.color_samples
    end
  end

  def test_grey_alias
    assert_equal ColorizedString['This is grey'].grey, ColorizedString['This is grey'].light_black
  end

  def test_gray_alias
    assert_equal ColorizedString['This is gray'].gray, ColorizedString['This is gray'].light_black
  end

  def test_add_color_alias
    ColorizedString.add_color_alias(:extra_blue, :light_blue)

    assert_equal ColorizedString['blue'].light_blue, ColorizedString['blue'].extra_blue
    assert_equal ColorizedString['blue'].on_light_blue, ColorizedString['blue'].on_extra_blue

    assert_raises ::Colorize::ColorAlreadyExist, 'Colorize: color :extra_blue already exist!' do
      ColorizedString.add_color_alias(:extra_blue, :light_color)
    end

    assert_raises ::Colorize::ColorDontExist, 'Colorize: color :light_color don\'t exist!' do
      ColorizedString.add_color_alias(:extra_white, :light_color)
    end
  end

  def test_add_color_alias_with_single_hash
    ColorizedString.add_color_alias(extra_green: :light_green)

    assert_equal ColorizedString['example string'].light_green, ColorizedString['example string'].extra_green
    assert_equal ColorizedString['example string'].on_light_green, ColorizedString['example string'].on_extra_green
  end

  def test_add_color_alias_with_single_hash_with_arrow
    ColorizedString.add_color_alias(:extra_color => :gray)

    assert_equal ColorizedString['example string'].gray, ColorizedString['example string'].extra_color
    assert_equal ColorizedString['example string'].on_gray, ColorizedString['example string'].on_extra_color
  end

  def test_add_color_alias_with_multi_hash
    ColorizedString.add_color_alias(extra_color_a: :gray, extra_color_b: :blue)

    assert_equal ColorizedString['example string'].gray, ColorizedString['example string'].extra_color_a
    assert_equal ColorizedString['example string'].blue, ColorizedString['example string'].extra_color_b
  end

  def test_add_color_alias_with_multi_hash_with_arrow
    ColorizedString.add_color_alias(:extra_color_c => :gray, :extra_color_d => :blue)

    assert_equal ColorizedString['example string'].gray, ColorizedString['example string'].extra_color_c
    assert_equal ColorizedString['example string'].on_blue, ColorizedString['example string'].on_extra_color_d
  end

  def test_add_color_alias_with_multi_hash_mixed
    ColorizedString.add_color_alias(extra_color_e: :gray, :extra_color_f => :blue)

    assert_equal ColorizedString['example string'].gray, ColorizedString['example string'].extra_color_e
    assert_equal ColorizedString['example string'].on_blue, ColorizedString['example string'].on_extra_color_f
  end

  def test_prevent_colors
    ColorizedString.prevent_colors true

    assert ColorizedString.prevent_colors
    assert_equal "#{ColorizedString['blue'].blue}#{ColorizedString['red'].red}#{ColorizedString['green'].green}", ColorizedString["#{ColorizedString['blue'].blue}red#{ColorizedString['green'].green}"].red

    ColorizedString.prevent_colors false
  end
end


lib/colorize/class_methods.rb
# frozen_string_literal: true

module Colorize
  module ClassMethods
    @@color_codes ||= {
      black:        30,
      red:          31,
      green:        32,
      yellow:       33,
      blue:         34,
      purple:       35,
      cyan:         36,
      white:        37,
      default:      39,
      light_black:  90,
      light_red:    91,
      light_green:  92,
      light_yellow: 93,
      light_blue:   94,
      light_purple: 95,
      light_cyan:   96,
      light_white:  97
    }

    @@mode_codes ||= {
      default:          0, # Turn off all attributes
      bold:             1,
      dim:              2,
      italic:           3,
      underline:        4,
      blink:            5,
      blink_slow:       5,
      blink_fast:       6,
      invert:           7,
      hide:             8,
      strike:           9,
      double_underline: 20,
      reveal:           28,
      overlined:        53
    }

    #
    # Property for disable colorization
    #
    def disable_colorization(value = nil)
      if value.nil?
        if defined?(@@disable_colorization)
          @@disable_colorization || false
        else
          false
        end
      else
        @@disable_colorization = (value || false)
      end
    end

    #
    # Setter for disable colorization
    #
    def disable_colorization=(value)
      @@disable_colorization = (value || false)
    end

    #
    # Property for prevent recolorization
    #
    def prevent_colors(value = nil)
      if value.nil?
        if defined?(@@prevent_colors)
          @@prevent_colors || false
        else
          false
        end
      else
        @@prevent_colors = (value || false)
      end
    end

    #
    # Setter for prevent recolorization
    #
    def prevent_colors=(value)
      @@prevent_colors = (value || false)
    end

    #
    # Return array of available colors used by colorize
    #
    def colors
      color_codes.keys
    end

    #
    # Return array of available modes used by colorize
    #
    def modes
      mode_codes.keys
    end

    #
    # Display color samples
    #
    def color_samples
      colors.permutation(2).each do |background, color|
        sample_text = "#{color.inspect.rjust(15)} on #{background.inspect.ljust(15)}"
        puts "#{new(sample_text).colorize(:color => color, :background => background)} #{sample_text}"
      end
    end

    #
    # Add color alias
    #
    def add_color_alias(*params)
      parse_color_alias_params(params).each do |_alias_, _color_|
        check_if_color_available!(_alias_)
        check_if_color_exist!(_color_)

        add_color_code(_alias_, color_codes[_color_])
        add_color_method(_alias_)
      end
    end

    # private

    #
    # Color codes hash
    #
    def color_codes
      @@color_codes
    end

    def add_color_code(code, color)
      @@color_codes[code] = color
    end

    #
    # Mode codes hash
    #
    def mode_codes
      @@mode_codes
    end

    #
    # Generate color and on_color methods
    #
    def color_methods
      colors.each do |key|
        next if key == :default

        add_color_method(key)
      end
    end

    #
    # Generate color and on_color method
    #
    def add_color_method(key)
      define_method key do
        colorize(:color => key)
      end

      define_method "on_#{key}" do
        colorize(:background => key)
      end
    end

    #
    # Generate modes methods
    #
    def modes_methods
      modes.each do |key|
        next if key == :default

        define_method key do
          colorize(:mode => key)
        end
      end
    end

    def parse_color_alias_params(params)
      return [params] if params.is_a?(Array) && params.length == 2

      params.flat_map do |param|
        next param if param.is_a?(Array) && param.length == 2
        next param.to_a if param.is_a?(Hash)
      end
    end

    #
    # Check if color exists
    #
    def check_if_color_available!(color)
      color_codes[color] && fail(::Colorize::ColorAlreadyExist, "Colorize: color named :#{color} already exist!")
    end

    #
    # Check if color is missing
    #
    def check_if_color_exist!(color)
      color_codes[color] || fail(::Colorize::ColorDontExist, "Colorize: color :#{color} don't exist!")
    end
  end
end


lib/colorize/instance_methods.rb
# frozen_string_literal: true

module Colorize
  module InstanceMethods
    #
    # Change color of string
    #
    # Examples:
    #
    #   puts "This is blue".colorize(:blue)
    #   puts "This is light blue".colorize(:light_blue)
    #   puts "This is also blue".colorize(:color => :blue)
    #   puts "This is light blue with red background".colorize(:color => :light_blue, :background => :red)
    #   puts "This is light blue with red background".colorize(:light_blue ).colorize( :background => :red)
    #   puts "This is blue text on red".blue.on_red
    #   puts "This is red on blue".colorize(:red).on_blue
    #   puts "This is red on blue and underline".colorize(:red).on_blue.underline
    #   puts "This is blue text on red".blue.on_red.blink
    #   puts "This is uncolorized".blue.on_red.uncolorize
    #
    def colorize(params)
      return self if self.class.disable_colorization

      scan_for_colors.inject(self.class.new) do |str, match|
        colors_from_params(match, params)
        defaults_colors(match)
        str << "\001\033[#{match[0]};#{match[1]};#{match[2]}m\002#{match[3]}\001\033[0m\002"
      end
    end

    #
    # Return uncolorized string
    #
    def uncolorize
      scan_for_colors.inject(self.class.new) do |str, match|
        str << match[3]
      end
    end

    #
    # Return true if string is colorized
    #
    def colorized?
      scan_for_colors.inject([]) do |colors, match|
        colors << match.tap(&:pop)
      end.flatten.compact.any?
    end

    private

    #
    # Set default colors
    #
    def defaults_colors(match)
      match[0] ||= mode(:default)
      match[1] ||= color(:default)
      match[2] ||= background_color(:default)
    end

    #
    # Set color from params
    #
    def colors_from_params(match, params)
      case params
      when Hash then colors_from_hash(match, params)
      when Symbol then color_from_symbol(match, params)
      end
    end

    #
    # Set colors from params hash
    #
    def colors_from_hash(match, hash)
      if self.class.prevent_colors
        match[0] ||= mode(hash[:mode])                   if mode(hash[:mode])
        match[1] ||= color(hash[:color])                 if color(hash[:color])
        match[2] ||= background_color(hash[:background]) if background_color(hash[:background])
      else
        match[0] = mode(hash[:mode])                   if mode(hash[:mode])
        match[1] = color(hash[:color])                 if color(hash[:color])
        match[2] = background_color(hash[:background]) if background_color(hash[:background])
      end
    end

    #
    # Set color from params symbol
    #
    def color_from_symbol(match, symbol)
      if self.class.prevent_colors && color(symbol)
        match[1] ||= color(symbol)
      else
        match[1] = color(symbol)
      end
    end

    #
    # Color for foreground
    #
    def color(color)
      self.class.color_codes[color]
    end

    #
    # Color for background (offset 10)
    #
    def background_color(color)
      self.class.color_codes[color] + 10 if self.class.color_codes[color]
    end

    #
    # Mode
    #
    def mode(mode)
      self.class.mode_codes[mode]
    end

    #
    # Scan for colorized string
    #
    def scan_for_colors
      scan(/\001\033\[([0-9;]+)m\002(.+?)\001\033\[0m\002|([^\001\033]+)/m).map do |match|
        split_colors(match)
      end
    end

    def split_colors(match)
      colors = (match[0] || '').split(';')
      array = Array.new(3)
      array[0], array[1], array[2] = colors if colors.length == 3
      array[1] = colors                     if colors.length == 1
      array[3] = match[1] || match[2]
      array
    end

    #
    # Require windows libs
    #
    def require_windows_libs
      require 'Win32/Console/ANSI' if RUBY_VERSION < '2.0.0' && RUBY_PLATFORM.include?('win32')
    rescue LoadError
      raise 'You must gem install win32console to use colorize on Windows'
    end
  end
end
